app/services/optout/OptOutService.scala

```diff
@@ -146,16 +148,31 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
     setupOptOutProposition().map { proposition =>
       proposition.optOutPropositionType.flatMap {
         case p: OneYearOptOutProposition => Some(ConfirmedOptOutViewModel(optOutTaxYear = p.intent, state = p.state()))
-        case p: MultiYearOptOutProposition => intent.map(i => ConfirmedOptOutViewModel(optOutTaxYear = i, state = p.state()))
+        case p: MultiYearOptOutProposition => intent.map(i => ConfirmedOptOutViewModel(optOutTaxYear = i, state = p.state()))//todo: add test code
       }
     }
   }
 
+
+  def getTaxYearsAvailableForOptOut()(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Seq[TaxYear]] = {
+    setupOptOutProposition().map(proposition => proposition.availableTaxYearsForOptOut)
+  }
+
+  def getSubmissionCountForTaxYear(offeredOptOutYears: Seq[TaxYear])
+                                  (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Map[Int, Int]] = {
+    val futureCounts: Seq[Future[(Int, Int)]] = offeredOptOutYears.map { optOutTaxYear =>
+      nextUpdatesService.getSubmissionCounts(optOutTaxYear)
+    }
+
+    Future.sequence(futureCounts).map(_.toMap)
+  }
+
 }
 
 
 object OptOutService {
-  def combineByReturningAnyFailureFirstOrAnySuccess(responses: Seq[Future[OptOutUpdateResponse]])(implicit ec: ExecutionContext): Future[OptOutUpdateResponse] = {
+
+  private def combineByReturningAnyFailureFirstOrAnySuccess(responses: Seq[Future[OptOutUpdateResponse]])(implicit ec: ExecutionContext): Future[OptOutUpdateResponse] = {
```

Comment:
* passing future into another function is a bit dangerous practice, this was mentioned few times in the past, if this can be avoided pls; Check existing code base for examples / ask questions if needed;

Reply:
Thank you Pavel, I accept what you say and will try to change it but I just don't understand it, for my own learning can you please explain the the danger here? or share a web link that explains the problem with this approach? also I'm not aware of previous discussions within the team about this as I still consider myself to be a recent joiner, this would be appreciated, thanks again

Reply:
* any variable referring to a type of Future is an operation in progress, this would not be possible clearly reason about code/function (we can disregard fact that its a private in this context ) which processing an operation as an input parameter; any Future we have we expect to compose with other Future or transform  again to yet another Future - this is yet the only way we can provide guarantees what we have and can easy read code produced;
*
* when you designing with any function returning Future[_], you can ask yourself if you can split this function into synchronous and asynchronous part? this simplify understanding what you actually do and actual expectation;
** - Scala Future is not an easy subject and there are lots of controversial materials online regarding  how to use these, I would suggest to refer to actual source code + raise question in the relevant channels we have;

Reply:
thanks for that, happy to pair with you to change it, otherwise I can't I see anything worng with the code and hance can't think how to make it less "dangerous"

Reply:
done, improved it a bit by moving the methods locally to the called public method so that it's all within the same single method, will look at it agian to improve it even further, thanks